database.hh

```diff
@@ -1073,6 +1076,13 @@ private:
 public:
     void update_off_strategy_trigger();
     void enable_off_strategy_trigger();
+
+private:
+    boost::icl::interval_map<dht::token, api::timestamp_type, boost::icl::partial_absorber, std::less, boost::icl::inplace_max> _repair_time_map;
```

Comment:
we need to repopulate map on boot by reading from system table no? so either a crash or restart will not prevent tombstones from being purged.

Reply:
Shouldn't we use `gc_clock::time_point` here? That is the clock that we use for expiry, the timestamp is a logical clock, which cannot be used to record the time of events.

Reply:
Yes, we need to repopulate on restart. As I mentioned in the doc, it is not done yet and is in the todo section.

Reply:
> Shouldn't we use gc_clock::time_point here? That is the clock that we use for expiry, the timestamp is a logical clock, which cannot be used to record the time of events.

Here, I wanted to record when the repair happened, like when the tombstone is written. When to expire is not related.  Why can't we use api::timestamp_type to reocrd the timestamp of an event.

Reply:
You can't mix timestamps with expiry clock, they are two different clocks. Tombstone expiry uses `gc_clock`, if you want to compare a timestamp to a given tombstone expiry (can I expire this already or not), you have to use `gc_clock`.
Also, like I said, `api::timestamp_type` cannot be used to record the timestamp of an event (like when last repair happened). It is a logical clock, kinda like a counter that you just use to order writes in some sequence, by their arrival to the db.

Reply:
We have

```
 35 struct tombstone final {
 36     api::timestamp_type timestamp;
 37     gc_clock::time_point deletion_time;
```

I am wondering how we decide the deletion_time. Do we convert the timestamp to deletion_time? 

We can also look at the repair as a "seqneunce"  just like read and write. We use the logical time to present the event.

Reply:
`deletion_time` is determined as `gc_clock::now()` for when the deletion happens. For expiring cells, this is the time the cell was written.

The problem with timestamps is that the user is allowed to override them, so they can create a tombstone that will never be expired if you use timestamp for when repair happened. This is also the reason we have a separate `gc_clock`.